test/mip/bound_strengthening.jl

```diff
@@ -0,0 +1,137 @@
+                x, y, z are integer  =#
+    C = T.([1, 1, 1])
+    lc = T.([-Inf, -Inf, -1])
+    uc = T.([Inf, Inf, 1])
+    lr = Vector{T}([-10])
+    ur = Vector{T}([10])
+    varTypes = [MOP.GENERAL_INTEGER, MOP.BINARY, MOP.GENERAL_INTEGER]
+    pb = MOP.ProblemData{T}()
+    A = sparse([1, 1], [1, 2], [10, -1], 1, 3)
+    MOP.load_problem!(pb, "Test",
+        true, C, zero(T),
+        A, lr, ur, lc, uc,
+        varTypes
+    )
+
+    ps = MOP.PresolveData(pb)
+    MOP.bound_strengthening!(ps)
+    @test ps.lcol == T.([-Inf, -Inf, -1])
+    @test ps.ucol == T.([Inf, Inf, 1])
+
+    # Build the following model
+    #=
+        min     x + y + z
+        s.t.    -10 ⩽ 10 x - y + 0 z ⩽ 10
+                x, y, z are integer  =#
+    C = T.([1, 1, 1])
+    lc = T.([-Inf, -Inf, -1])
+    uc = T.([Inf, Inf, 1])
+    lr = Vector{T}([-10])
+    ur = Vector{T}([10])
+    varTypes = [MOP.GENERAL_INTEGER, MOP.BINARY, MOP.GENERAL_INTEGER]
+    pb = MOP.ProblemData{T}()
+    A = sparse([1, 1], [1, 2], [10, -1], 1, 3)
+    MOP.load_problem!(pb, "Test",
+        true, C, zero(T),
+        A, lr, ur, lc, uc,
+        varTypes
+    )
+
+    ps = MOP.PresolveData(pb)
+    MOP.bound_strengthening!(ps)
+    @test ps.lcol == T.([-Inf, -Inf, -1])
+    @test ps.ucol == T.([Inf, Inf, 1])
+
+    # Build the following model
+    #=
+        min     x + y + z
+        s.t.    0 ⩽ x + y - z ⩽ 1
+                -1 ⩽ x ⩽ 1
+                x, y, z are integer  =#
+    C = T.([1, 1, 1])
+    lc = T.([-1, -Inf, -Inf])
+    uc = T.([1, Inf, Inf])
+    lr = Vector{T}([0])
+    ur = Vector{T}([1])
+    varTypes = [MOP.GENERAL_INTEGER, MOP.GENERAL_INTEGER, MOP.GENERAL_INTEGER]
+    pb = MOP.ProblemData{T}()
+    A = sparse([1, 1, 1], [1, 2, 3], [1, 1, -1], 1, 3)
+    MOP.load_problem!(pb, "Test",
+        true, C, zero(T),
+        A, lr, ur, lc, uc,
+        varTypes
+    )
+
+    ps = MOP.PresolveData(pb)
+    MOP.bound_strengthening!(ps)
+    @test ps.lcol == T.([-1, -Inf, -Inf])
+    @test ps.ucol == T.([1, Inf, Inf])
+    # Build the following model
+    #=
+        min     x + y + z
+        s.t.    MOP.eps(T)/2 ⩽ x + z ⩽ MOP.eps(T)/2
+                -MOP.eps(T)/2 ⩽ y + z ⩽ -MOP.eps(T)/2
+                MOP.eps(T)/2 ⩽ x + y ⩽ -MOP.eps(T)/2
+                0 ⩽ x ⩽ 1
+                x, y, z are integer  =#
+    C = T.([1, 1, 1])
+    lc = Vector{T}([0, -Inf, -Inf])
+    uc = Vector{T}([1, Inf, Inf])
+    lr = T.([MOP.eps(T)/2, -MOP.eps(T)/2, MOP.eps(T)/2])
+    ur = T.([MOP.eps(T)/2, -MOP.eps(T)/2, -MOP.eps(T)/2])
+
+    varTypes = [MOP.GENERAL_INTEGER, MOP.GENERAL_INTEGER, MOP.GENERAL_INTEGER]
+    pb = MOP.ProblemData{T}()
+    A = sparse([1, 1, 2, 2, 3, 3], [1, 3, 2, 3, 1, 2], [1, 1, 1, 1, 1, 1], 3, 3)
+    MOP.load_problem!(pb, "Test",
+        true, C, zero(T),
+        A, lr, ur, lc, uc,
+        varTypes
+    )
+
+    ps = MOP.PresolveData(pb)
+    MOP.bound_strengthening!(ps)
+    @test ps.lcol == T.([0, -1, 0])
+    @test ps.ucol == T.([1, 0, 0])
```

Comment:
I'm a bit confused how a single pass of the procedure is able to infer that z >= 0. Just looking constraint-wise, you should only be able to get -1 <= z <= 0 from the first row (with the original bounds). I guess it's because ``bound_strengthening`` first tightens bounds for y to [-1,0], and then proceeds to z, you can then infer z >= 0?

If the behavior of ``bound_strengthening!`` depends on the ordering of the variables, this should be documented somewhere IMO.

Reply:
In the single pass of the procedure, the bounds of x won't be updated since y and z are unbounded. Then, the bounds of y are updated to [-1, 0] since 0 <= x+y <= 0. After that, since 0 <= x+z <= 0, 0 <= y+z <= 0, x in [0, 1], and y in [-1, 0], z can only be 0.